src/Analyzer/Resolve/QueryAnalyzer.cpp

```diff
@@ -3102,10 +3145,127 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
                     in_second_argument = in_second_argument->cloneAndReplace(table_expression, std::move(replacement_table_expression_table_node));
                 }
             }
+            /// If it's a function node like array(..) or tuple(..), consider rewriting them to 'has':
+            if (auto * non_const_set_candidate = in_second_argument->as<FunctionNode>())
+            {
+                const auto & candidate_name = non_const_set_candidate->getFunctionName();
+
+                /// Case 1: array(..) node – keep existing behavior
+                if (candidate_name == "array")
+                {
+                    bool contains_nullable = false;
+                    for (const auto & array_elem : non_const_set_candidate->getArguments().getNodes())
+                    {
+                        if (isNullableOrLowCardinalityNullable(array_elem->getResultType()))
+                        {
+                            contains_nullable = true;
+                            break;
+                        }
+                    }
+                    if (!contains_nullable)
+                    {
+                        function_name = "has";
+                        is_special_function_in = false;
+                        auto & fn_args = function_node.getArguments().getNodes();
+                        std::swap(fn_args[0], fn_args[1]);
+                    }
```

Comment:
I think we need the `else` part with `makeNullSafeHas` here.
And check `transform_null_in`